driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -12,10 +16,52 @@ use futures::{
 use orderbook::Orderbook;
 use std::sync::{Arc, Mutex};
 
-// TODO:
-// `pub struct UpdatingOrderbook` that also owns an `Arc<Mutex<Orderbook>>` and creates a
-// background thread running `UpdatingOrderbookThread::update_with_events`.
-// This way `UpdatingOrderbookThread` is independently testable.
+/// An event based orderbook that automatically updates itself with new events from the contract.
+#[derive(Debug)]
+pub struct UpdatingOrderbook {
+    orderbook: Arc<Mutex<Orderbook>>,
+    // When this struct is dropped this sender will be dropped which makes the updater thread stop.
+    _channel: oneshot::Sender<()>,
+}
+
+impl UpdatingOrderbook {
+    pub fn new(
+        _contract: &impl StableXContract,
+        block_timestamp: impl BlockTimestamp + Send + 'static,
+    ) -> Self {
+        let orderbook = Arc::new(Mutex::new(Orderbook::default()));
+        let orderbook_clone = orderbook.clone();
+        let (sender, receiver) = oneshot::channel();
+        // Create stream first to make sure we do not miss any events between it and past events.
+        // TODO: use the real functions once they are implemented
+        let stream = futures::stream::iter(vec![]).boxed(); // contract.stream_events();
```

Comment:
Note that just creating the stream is not enough - it needs to be polled (and possibly `select!`ed with the `past_events` and `receiver` futures) so that it actually creates the filter on the node and starts polling it.

Reply:
Hmm and we need to make sure GetFilterChanges is called constantly and web3 probably has an internal buffer of events and stops calling that if the buffer fills up which could make the filter get deleted.
We could as you say select! it with the past events and forward the stream events to the orderbook too as it doesn't matter if they come in the wrong order.
Then with regards to your comment below there wouldn't be two parts but we could keep a flag "orderbook finished catching up" that is set when past events are handled and querying the orderbook before that fails with an error. Then we can also lock the mutex for all past events because flag would get checked before locking the mutex.